### ISSUE TITLE
Refactor JavadocLeadingAsteriskAlign to use logging of violation on AST node

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#20896]
split from https://github.com/checkstyle/checkstyle/issues/5770

JavadocLeadingAsteriskAlign should be updated to use ` public final void log(DetailAST ast,`
instead of
`public final void log(int lineNo, int colNo,`

Field `expectedColumnNumberWithoutExpandedTabs` should be investigated we should be able to avoid usage of it.
It was created long time ago, we changed AST last summer, so we should be able to work without it.

**Comment by arimu1** [comment]
In #20904 I moved the closing-tag (`*/`) violation over to `log(DetailAST ast, ...)`, but I got stuck on the mid-comment leading-asterisk violation in `visitJavadocToken`.

The problem: `LEADING_ASTERISK` in the javadoc lexer grammar is `[ \t]* '*' {isAfterNewline()}?`, so the token itself swallows the line's leading whitespace and always starts at column 1. If I log against that `DetailNode` (via `AbstractJavadocCheck#log(DetailNode, ...)`), every reported column collapses to 1 instead of the asterisk's real column - it breaks all 15 expected violations in the test.

So finishing this needs a decision on how the column should be derived, not just a mechanical swap. A couple of options I can think of:
- keep `log(int, int, ...)` here specifically (maybe with a documented/annotated exception), or
- compute the real column from the raw source line at logging time and pass a synthetic value some other way, or
- change what `LEADING_ASTERISK` matches in the grammar so the token itself starts at the asterisk.

Which direction would you prefer? Happy to do the follow-up once I know the shape you want.

**Comment by romani** [comment]
@gianmarcoschifone , please help here

**Comment by gianmarcoschifone** [comment]
@arimu1 

Option 3 looks like the cleanest solution if it can be done without regressions.
But after #20839, leading asterisks are also part of the block-tag parsing behavior. The lexer now accepts one or more leading asterisks and the AST can expose multiple-prefix cases as LEADING_ASTERISKS. So changing LEADING_ASTERISK to start at `*` may require additional lexer/parser work to keep the new block-tag behavior intact, especially around indentation before the leading stars.
So I would prefer option 3 only if it stays reasonably scoped. Otherwise I am fine with option 2 keeping the current token shape and derive the reported column from the node text using something like `node.getColumnNumber() + node.getText().indexOf('*')` for LEADING_ASTERISK / LEADING_ASTERISKS.

**Comment by romani** [comment]
@arimu1 , I need your help to let me understand the problem.
Please share java file that can reproduce this problem to let me see problem.

**Comment by arimu1** [comment]
@romani here is a minimal file that shows the mid-comment column problem.

```java
/*xml
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocLeadingAsteriskAlign"/>
  </module>
</module>
*/
package demo;

public class ReproLeadingAsteriskColumn {
  /**
   * Opening asterisk is at column 4 (0-based: 3).
  * This mid-comment asterisk is also at column 4 — correctly aligned.
    * This mid-comment asterisk is at column 6 — misaligned (expected 4).
   */
  void example() {}
}
```

**What goes wrong if we switch the mid-comment path to `log(DetailNode)`**

Lexer rule (`JavadocCommentsLexer.g4`):

```
LEADING_ASTERISK : [ \t]* '*' {isAfterNewline()}? -> channel(LEADING_ASTERISKS)
```

So the token text for the misaligned line is `"    *"` (spaces + star), and the token **starts at column 1** of that javadoc line. `AbstractJavadocCheck.log(DetailNode)` does:

```java
log(node.getLineNumber(), node.getColumnNumber(), key, args);
```

…so every mid-comment leading-asterisk violation would report **column 1**, even when the `*` itself sits at column 6.

Today `visitJavadocToken` avoids that by deriving the real `*` column from `ast.getText()` (`getAsteriskColumnNumber`) and calling `log(line, column, …)` with that value. That is why `InputJavadocLeadingAsteriskAlignIncorrect.java` expects things like `19:5`, `26:3`, `43:7` — those are the `*` columns, not the token start.

The closing `*/` path in #20904 was able to use `log(DetailAST)` because the Java block-comment AST node for `*/` already sits on the asterisk. Mid-comment `LEADING_ASTERISK` does not.

**Same shape already lives in the tree as**  
`src/test/resources/.../javadocleadingasteriskalign/InputJavadocLeadingAsteriskAlignIncorrect.java`  
(e.g. the `foo3` / `foo4` cases).

Happy to implement @gianmarcoschifone's option 2 next (`node.getColumnNumber() + node.getText().indexOf('*')` for `LEADING_ASTERISK` / `LEADING_ASTERISKS`, keep current token shape) unless you want a different direction after looking at this.

**Comment by romani** [comment]
```
$ cat Test.javadoc 
   * Opening asterisk is at column 4 (0-based: 3).
  * This mid-comment asterisk is also at column 4 — correctly aligned.
    * This mid-comment asterisk is at column 6 — misaligned (expected 4).

$ java -jar checkstyle-13.9.0-SNAPSHOT-all.jar -j Test.javadoc 
JAVADOC_CONTENT -> JAVADOC_CONTENT [0:1]
|--LEADING_ASTERISK ->    * [0:1]
|--TEXT ->  Opening asterisk is at column 4 (0-based: 3). [0:5]
|--NEWLINE -> \n [0:51]
|--LEADING_ASTERISK ->   * [1:1]
|--TEXT ->  This mid-comment asterisk is also at column 4 — correctly aligned. [1:4]
|--NEWLINE -> \n [1:71]
|--LEADING_ASTERISK ->     * [2:1]
`--TEXT ->  This mid-comment asterisk is at column 6 — misaligned (expected 4). [2:6]

$ cat Test.java
/** some. */
public class Test {
  /**
   * Opening asterisk is at column 4 (0-based: 3).
  * This mid-comment asterisk is also at column 4 — correctly aligned.
    * This mid-comment asterisk is at column 6 — misaligned (expected 4).
   */
  void example() {}
}

$ java -jar checkstyle-13.9.0-SNAPSHOT-all.jar -J Test.java 
COMPILATION_UNIT -> COMPILATION_UNIT [2:1]
`--CLASS_DEF -> CLASS_DEF [2:1]
    |--MODIFIERS -> MODIFIERS [2:1]
    |   |--BLOCK_COMMENT_BEGIN -> /* [1:1]
    |   |   |--COMMENT_CONTENT -> * some.  [1:3]
    |   |   |   `--JAVADOC_CONTENT -> JAVADOC_CONTENT [1:4]
    |   |   |       `--TEXT ->  some.  [1:4]
    |   |   `--BLOCK_COMMENT_END -> */ [1:10]
    |   `--LITERAL_PUBLIC -> public [2:1]
    |--LITERAL_CLASS -> class [2:8]
    |--IDENT -> Test [2:14]
    `--OBJBLOCK -> OBJBLOCK [2:19]
        |--LCURLY -> { [2:19]
        |--METHOD_DEF -> METHOD_DEF [8:3]
        |   |--MODIFIERS -> MODIFIERS [8:3]
        |   |--TYPE -> TYPE [8:3]
        |   |   |--BLOCK_COMMENT_BEGIN -> /* [3:3]
        |   |   |   |--COMMENT_CONTENT -> *\n   * Opening asterisk is at column 4 (0-based: 3).\n  * This mid-comment asterisk is also at column 4 — correctly aligned.\n    * This mid-comment asterisk is at column 6 — misaligned (expected 4).\n    [3:5]
        |   |   |   |   `--JAVADOC_CONTENT -> JAVADOC_CONTENT [3:6]
        |   |   |   |       |--NEWLINE -> \n [3:6]
        |   |   |   |       |--LEADING_ASTERISK ->    * [4:1]
        |   |   |   |       |--TEXT ->  Opening asterisk is at column 4 (0-based: 3). [4:5]
        |   |   |   |       |--NEWLINE -> \n [4:51]
        |   |   |   |       |--LEADING_ASTERISK ->   * [5:1]
        |   |   |   |       |--TEXT ->  This mid-comment asterisk is also at column 4 — correctly aligned. [5:4]
        |   |   |   |       |--NEWLINE -> \n [5:71]
        |   |   |   |       |--LEADING_ASTERISK ->     * [6:1]
        |   |   |   |       |--TEXT ->  This mid-comment asterisk is at column 6 — misaligned (expected 4). [6:6]
        |   |   |   |       |--NEWLINE -> \n [6:74]
        |   |   |   |       `--TEXT ->     [7:1]
        |   |   |   `--BLOCK_COMMENT_END -> */ [7:4]
        |   |   `--LITERAL_VOID -> void [8:3]
        |   |--IDENT -> example [8:8]
        |   |--LPAREN -> ( [8:15]
        |   |--PARAMETERS -> PARAMETERS [8:16]
        |   |--RPAREN -> ) [8:16]
        |   `--SLIST -> { [8:18]
        |       `--RCURLY -> } [8:19]
        `--RCURLY -> } [9:1]
```

`|--LEADING_ASTERISK ->   * [1:1]` this is wrong column. Line 1 has `*` at column 4, Line 2 has `*` at column 3, Line 3 has `*` at column 5.

Ideally it should be:
```
|--NEWLINE -> \n [3:6]
|--WS ->     [4:1]
|--LEADING_ASTERISK -> * [4:4]
|--TEXT ->  Opening asterisk is at column 4 (0-based: 3). [4:5]
```

**Comment by romani** [comment]
@arimu1 , please continue to upgrade Check to report issue on token, position is wrong, it is unfortunate. but it is problem of parser.
As we fix parser, all become back to good.

@gianmarcoschifone , lets continue discussion on parser problem at https://github.com/checkstyle/checkstyle/issues/20922

**Comment by romani** [comment]
@arimu1 , we merged update for parser, migration of Check to AST should be smooth now , please continue.

**Comment by raunakdeysarkar** [comment]
Hi @romani, since the parser update in #20922 is merged, I'd like to work on migrating JavadocLeadingAsteriskAlignCheck to use AST node logging. Could you please assign this issue to me?

**Comment by AliAkbarBaloch** [comment]
> Hi [@romani](https://github.com/romani), since the parser update in [#20922](https://github.com/checkstyle/checkstyle/issues/20922) is merged, I'd like to work on migrating JavadocLeadingAsteriskAlignCheck to use AST node logging. Could you please assign this issue to me?

He don't assign issues. If the issue hasn't been claimed yet, please comment on it before you start working. Once your work is complete & tested , everything is good then create a pull request.

**Comment by romani** [comment]
if no PR created or it is abandoned, make comment "I am on it" and start working.

**Comment by SidheshwarSarangal** [comment]
I am on it — I will migrate JavadocLeadingAsteriskAlignCheck to AST-node logging and investigate removal of expectedColumnNumberWithoutExpandedTabs.

**Comment by Youssefalaa7** [comment]
I am on it
